File: content/bdd-with-cucumber/07/code/js/07-shout-a-message-everywhere/src/shouty.js
class Person {
  constructor(network, location) {
    this.messages = []
    this.credits = 0
    this.network = network
    this.location = location

    this.network.subscribe(this)
  }

  shout(message) {
    this.network.broadcast(message, this)
  }

  hear(message) {
    this.messages.push(message)
  }

  messagesHeard() {
    return this.messages
  }
}

class Network {
  constructor(range) {
    this.listeners = []
    this.range = range
  }

  subscribe(person) {
    this.listeners.push(person)
  }

  broadcast(message, shouter) {
    const shortEnough = message.length <= 180
    this._deductCredits(shortEnough, message, shouter)
    this.listeners.forEach((listener) => {
      const withinRange =
        Math.abs(listener.location - shouter.location) <= this.range
      if (withinRange && (shortEnough || shouter.credits >= 0)) {
        listener.hear(message)
      }
    })
  }

  _deductCredits(shortEnough, message, shouter) {
    if (!shortEnough) shouter.credits -= 2
    shouter.credits -= (message.match(/buy/gi) || []).length * 5
  }
}

module.exports = {
  Person: Person,
  Network: Network,
}

File: content/bdd-with-cucumber/07/code/ruby/16-merge-repeated-over-long-steps/lib/shouty.rb
module Shouty
  class Person
    attr_reader :messages_heard, :location
    attr_accessor :credits

    def initialize(network, location)
      @messages_heard = []
      @network        = network
      @location       = location
      @credits        = 0

      @network.subscribe(self)
    end

    def shout(message)
      @network.broadcast(message, self)
    end

    def hear(message)
      @messages_heard << message
    end
  end

  class Network
    def initialize(range)
      @range     = range
      @listeners = []
    end

    def subscribe(person)
      @listeners << person
    end

    def broadcast(message, shouter)
      short_enough = message.size <= 180
      deduct_credits(short_enough, message, shouter)
      @listeners.each do |listener|
        within_range = (listener.location - shouter.location).abs <= @range

        if within_range && (short_enough || shouter.credits >= 0)
          listener.hear(message)
        end
      end
    end

    def deduct_credits(short_enough, message, shouter)
      shouter.credits -= 2 unless short_enough
      shouter.credits -= (message.scan(/buy/i) || []).size * 5
    end
  end
end
